internal/network/cidr_validations.go

```diff
@@ -119,33 +120,59 @@ func VerifyClusterCidrSize(hostNetworkPrefix int, clusterNetworkCIDR string, num
 	// 63 to avoid overflow
 	possibleNumHosts := uint64(1) << min(63, max(hostNetworkPrefix-clusterNetworkPrefix, 0))
 	if uint64(requestedNumHosts) > possibleNumHosts {
-		return errors.Errorf("Cluster network CIDR prefix %d does not contain enough addresses for %d hosts each one with %d prefix (%d addresses)",
+		return fmt.Errorf("Cluster network CIDR prefix %d does not contain enough addresses for %d hosts each one with %d prefix (%d addresses)",
 			clusterNetworkPrefix, numberOfHosts, hostNetworkPrefix, uint64(1)<<min(63, bits-hostNetworkPrefix))
 	}
 	return nil
 }
 
-func VerifyClusterCIDRsNotOverlap(machineNetworkCidr, clusterNetworkCidr, serviceNetworkCidr string, machineNetworkRequired bool) error {
-	if machineNetworkRequired {
-		err := VerifyNetworksNotOverlap(machineNetworkCidr, serviceNetworkCidr)
-		if err != nil {
-			return errors.Wrap(err, "MachineNetworkCIDR and ServiceNetworkCIDR")
+func VerifyNoNetworkCidrOverlaps(
+	clusterNetworks []*models.ClusterNetwork,
+	machineNetworks []*models.MachineNetwork,
+	serviceNetworks []*models.ServiceNetwork) error {
+	errs := []error{}
+	for imn, mn := range machineNetworks {
```

Comment:
High time complexity here O(m*n)
There might be something clever we could do not reduce this. 

Different language I know, but there are approaches using a data structure to memorize already searched ranges and to allow iteration across a set of networks in Order(n)

Note the use of a Radix tree in this implementation.

https://github.com/fgiuba/ipconflict/blob/master/ipconflict/subnet.py#L46-L61

Due to the limited number of networks, the consequences of not handling this may not be problematic. But it does look like something could be improved here!